### DISCUSSION TITLE
Problem in search of a proposal : avoid copying from single to multidim array when memory layout is compatible

### DISCUSSION THREAD

**Discussion by ebfortin** [dotnet/csharplang#3053] [Q&A]
I have two framework in which I have an array to manage storage for data. The data layer uses a Matrix class with a single dim array as its backing storage. The other framework, for data visualization, uses a multidim array as its input. The only way to transfer the data is with a copy, which is a shame since internally the memory layout of the single and multidim array is exactly the same. Wouldn't it be useful to be able to fake a multidim array from a single dim one?

**Comment by IS4Code** [comment]
You would probably have more luck getting your framework to support a single-dimensional array (or Span<T>) than getting such a feature (that would require significant runtime support) to the language.

**Comment by theunrepentantgeek** [comment]
If your data visualisation accepts an abstraction (e.g. `I2DMatrix`) you could write your own projection that maps onto a one dimensional array.

**Comment by huoyaoyuan** [comment]
> which is a shame since internally the memory layout of the single and multidim array is exactly the same

No. Although their elements layout are same, their array headers are different.

> Wouldn't it be useful to be able to fake a multidim array from a single dim one?

CLR does have faking feature for array, like `string[] -> object[]` and `int[] <-> uint[]`. Unfortunately, single dim array provides exactly less information than multi dim one. Because the faking feature cannot carry any extra information, it can be only done in the opposite way (`T[,] -> T[]`).

**Comment by huoyaoyuan** [comment]
Additional notes:
Object faking has been very problematic in the type system, and only kept for compatibility.
It's always considered as bad design to require concrete type of input data, especially arrays and strings. That's why span and memory are introduced. Bringing **holes into type system** to optimize **badly designed apis** should be non-goal of the runtime.

**Comment by ebfortin** [comment]
> No. Although their elements layout are same, their array headers are different.

I agree that the header is different. But the data layout is the same. That's why I can BlockCopy a single dim to a multi dim array. Faking would also fake the header. It's not really different than having a subclass of double[,] that uses a double[] as its backing store. However I can't do that right now.

**Comment by huoyaoyuan** [comment]
> having a subclass of double[,]

The memory&header layout of arrays and strings are strictly known everywhere, to avoid performance issues. That's why they cannot be subclass'd.
